021-team-credential-managers/proposal.md

```diff
@@ -0,0 +1,128 @@
+# Summary
+
+Move credential management configuration to the team configuration level, and
+allow configuring multiple credential managers for each team.
+
+
+# Motivation
+
+Configuring a single credential manager is very limiting. Because there is only
+one point of authentication configured, each credential manager has to support
+some form of multi-tenancy. The current strategy is to encode team and pipeline
+names in the paths for the keys that are looked up, but this has many
+downsides:
+
+* This makes it impossible to share credentials between teams. Instead the
+  credential has to be duplicated under each team's path.
+
+  By moving credential manager config to each team we can instead leverage the
+  credential manager's access control to determine how credentials are shared
+  across teams (e.g. Vault policies).
+
+* With Vault, this makes it impossible to use any backend except `kv`, because
+  all keys live under the same path scheme, and different backends can't be
+  mounted under paths managed by other backends. This removes a lot of the
+  value of using Vault in the first place.
+
+  By eliminating the path enforcement you can now refer to different secret
+  backend mount points.
+
+* Some credential managers, e.g. Azure KeyVault, have very strict requirements
+  for key names (`[a-z\-]+`), effectively making scoping conventions impossible
+  to enforce.
+
+  By configuring at the team level, each team can point to their own KeyVault
+  or configure their own access control.
+
+* It would be nice to be able to leverage a specialized credential manager like
+  IAM/STS for some things (like the `s3` resource) and use Vault for everything
+  else. Right now you can only configure one credential manager, so this is
+  impossible.
+
+  By allowing teams to configure multiple credential managers, all credential
+  managers can be tried in order when looking up a given credential.
```

Comment:
I would be vary about trying credential managers in order; if that means naively looking up secrets. It's very easy to hit rate limits on e.g. AWS SSM Parameter store, and AWS Secrets Manager is charged per request - related issue: https://github.com/concourse/concourse/issues/2924

Reply:
Hmm. How about a prefix scheme instead? e.g. `((vault.foo.bar))`

We initially avoided explicit prefixes as it directly tied the pipeline config to credential management, vs. being able to supply the vars via `-v`/`-l`, but I guess it's kind of unavoidable here, and ends up being more clear anyway, especially as different credential managers may have different look-up schemes. You could also directly put those prefixes in your `-v`/`-l` values, so... 🤔 Maybe I'll just change it to that?

Reply:
It's a little less convenient, but I think it makes sense to avoid rate limits and unnecessary costs 🤷‍♂️ Could maybe be helped a bit if was possible to set a default credentials provider, so you only have to specify `<provider>.foo.bar` only if you are using the non-default provider?

Sidenote; since we want to support nested credential paths like `((foo/bar/baz))`, and fields `((foo.bar))` - maybe providers should be specified as e.g. `((vault://foo/bar.baz))` instead of `((vault.foo/bar.baz))`?